torch/nn/utils/clip_grad.py

```diff
@@ -24,6 +26,7 @@ def clip_grad_norm_(
         error_if_nonfinite (bool): if True, an error is thrown if the total
             norm of the gradients from :attr:`parameters` is ``nan``,
             ``inf``, or ``-inf``. Default: False (will switch to True in the future)
+        foreach (bool): use the faster foreach-based implementation
```

Comment:
Not necessarily a thing to change on this PR, but is there any reason to not have this be the default? Is it because foreach is usually slower on CPU as it falls back to the slow path?

Reply:
Hahaha I am currently working on making foreach the default for optimizers, but there's a bit more complexity there since we have an option for differentiable. Likewise, there are other backends like mps and rocm that don't support foreach.

I think a good plan is to make the foreach kwarg an Optional[bool], to give the user ability to choose True or False, and then for us, since we know it's definitely faster on cpu and on cuda to turn on foreach = True if foreach was not specified by the user + the tensors are on CPU or CUDA.

Reply:
Yes makes sense, I'll turn it on. But why Optional[bool] and not just a bool with default True?

Reply:
If tensors are on a non CPU/CUDA device, won't the foreach call just go to "slow" path (which would still be faster than python)? If so there isn't any issue with setting it to default True regardless of device. Can't test it on a MPS device on my side.

Reply:
Also since we split the tensors by device anyway we could run _foreach only on the cpu/cuda tensor groups and do the python for loop for the rest

Reply:
Since we split the tensors by device anyway we could run _foreach only on the cpu/cuda tensor groups and do the python for loop for the rest. See my commit https://github.com/pytorch/pytorch/pull/91846/commits/618a54a47aa53ccc65599ab0c292a66d04f46da4 . This commit also makes the non-foreach version slightly faster thanks to the per-device grouping.

Reply:
noice!